README
PlayGo activity for the OLPC

Authors: Gerard J. Cerchio, 
		www.circlesoft.com
	 Andrés Ambrois (aa)
		andresambrois@gmail.com
		
Copyright under the GNU GPL



NEWS
6

* Python3 version
  - Fix Collaboration in old builds (Andres Ambrois)
  - Adding languages (Pootle translation push)
  - Add support for gamepad keys (Andres Ambrois)
  - Port to GTK+3 (Cristian Garcia)
  - Fixing draw problem (Cristian Garcia)
  - Add license and repository metadata (Shivam Chadha)
  - translate.sugarlabs.org synchronisation (Yash Agrawal, Jui Pradhan)
  - Regenerate POT file (Yash Agrawal, Jui Pradhan)
  - Fix PyGIWarnings (Yash Agarwal)
  - Fix NameError (Rahul Bothra)
  - Port Telepathy static bindings to TelepathyGLib (Ibiam Chihurumnaya)
  - Flake8 fixes (Ibiam Chihurumnaya)
  - Update README (rdrsadhu, James Cameron)
  - Port to python3 (Srevin Saju)

5

4
  - Fixed problem with sugartubeconnection import on old bilds (needs testing)
  - Nate Ridderman: Change references to 'Whites' and 'Blacks' to 'White' and 'Black'
  - Allow GnuGo to pick up game state at any point in time, including resuming from Journal
  - Mark territories at game end with pretty crosses. (Thanks Nate!)
  - Minor code cleanups. 
  - Added licensing information to activity.info (GPLv2 or later)
2
* Almost complete rewrite by aa. New features: 
  - Cleaner code
  - Save and resume games from Journal
  - Different Board sizes
  - Restart game button
  - Better collaboration (turn enforcement, notifications)
  - Pass and Undo in hotseat (not shared) mode. 
  - Full Spanish translation

1
* New activity PlayGo written by Gerard J. Cerchio (www.circlesoft.com)



gtp.py
# -*- coding: utf-8 -*-
# Copyright 2007-2008 One Laptop Per Child
# Copyright 2008 Andrés Ambrois <andresambrois@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

from subprocess import Popen, PIPE
import logging

from sugar3.activity.activity import get_bundle_path
from os.path import exists, join, abspath
from os import pathsep, environ

logger = logging.getLogger('PlayGo')


def search_for_gnugo():
    paths = environ['PATH'].split(pathsep)
    for path in paths:
        if exists(join(path, 'gnugo')):
            return abspath(join(path, 'gnugo'))
    default_path = join(get_bundle_path(), 'bin', 'gnugo')
    if exists(default_path):
        return abspath(default_path)

    return False


class gnugo:

    ''' A wrapper for talking to gnugo over GTP '''

    def __init__(self, boardsize=19, handicap=0, komi=5.5, level=3):
        ''' Start the gnugo subprocess '''
        self.size = boardsize
        self.path = search_for_gnugo()

        if self.path:
            logger.debug('Found gnugo at %s', self.path)
            try:
                self.gnugo = Popen([
                    self.path,
                    '--mode', 'gtp', '--boardsize', str(boardsize),
                    '--handicap', str(handicap), '--komi', str(komi),
                    '--level', str(level)], stdout=PIPE, stdin=PIPE)
            except OSError as data:
                logger.error('Could not start gnugo subprocess: %s', data)
                raise
            else:
                logger.debug('Successfuly loaded gnugo!')
                self.stdin = self.gnugo.stdin
                self.stdout = self.gnugo.stdout
        else:
            logger.error('Could not find gnugo')

    def __del__(self):
        logger.debug('Closing gnugo')
        self.stdin.write('quit \n')
        self.stdin.flush()

    def _xy_to_coords(self, x, y):
        return dict(
            list(zip(list(range(25)),
                 'ABCDEFGHJKLMNOPQRSTUVWXYZ')))[x] + str(self.size - y)

    def _coords_to_xy(self, coords):
        return int(
            dict(list(zip('ABCDEFGHJKLMNOPQRSTUVWXYZ',
                          list(range(25)))))[coords[0]]),
        self.size - int(coords[1:])

    def short_to_long_colors(self, short_color):
        if short_color == 'B':
            return 'black'
        return 'white'

    def make_play(self, color, x, y):
        color = self.short_to_long_colors(color)
        self.stdin.write('play %s %s\n' % (color, self._xy_to_coords(x, y)))
        self.stdin.flush()
        logger.debug(
            'Sent play by %s at %s to gnugo', color, self._xy_to_coords(x, y))
        output = self.stdout.readline()
        self.stdout.readline()
        if output and output[0] == '?':
            return False
        return True

    def get_move(self, color):
        color = self.short_to_long_colors(color)
        self.stdin.write('kgs-genmove_cleanup %s\n' % color)
        self.stdin.flush()
        output = self.stdout.readline()
        self.stdout.readline()
        if output[0] == '?':
            # FIXME: Handle error
            return False
        elif output[2:] == 'PASS\n':
            return -1, -1
        logger.debug('Generated move %s', output[2:])
        return self._coords_to_xy(output[2:])

    def pass_move(self, color):
        color = self.short_to_long_colors(color)
        self.stdin.write('%s pass\n' % color)
        self.stdin.flush()
        self.stdout.readline()
        self.stdout.readline()

    def undo(self):
        self.stdin.write('undo\n')
        self.stdin.flush()
        self.stdout.readline()
        self.stdout.readline()

    def clear(self):
        self.stdin.write('clear_board\n')
        self.stdin.flush()
        self.stdout.readline()
        self.stdout.readline()

    def dump_board(self):
        self.stdin.write('showboard\n')
        self.stdin.flush()
        output = ''
        for i in range(0, self.size + 4):
            output = output + self.stdout.readline()
        return output


gogame.py
# -*- coding: utf-8 -*-
# Copyright 2007-2008 One Laptop Per Child
# Copyright 2007 Gerard J. Cerchio <www.circlesoft.com>
# Copyright 2008 Andrés Ambrois <andresambrois@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import logging
_logger = logging.getLogger('PlayGo')

from gettext import gettext as _


class GoGame:
    """ This class administrates a go board.
        It keeps track of the stones currently on the board
        in the dictionary self.status, and of the moves played
        so far in self.undostack

        It has methods to clear the board, play a stone, undo a move. """

    def __init__(self, boardSize=19):
        self.size = boardSize
        self.status = {}
        self.undostack = []
        self.score = {'B': 0, 'W': 0}
        _logger.setLevel(logging.DEBUG)

    def get_score(self):
        return self.score

    def increase_score(self, color):
        self.score[color] = self.score[color] + 1

    def neighbors(self, x):
        """ Returns the coordinates of the 4
            (resp. 3 resp. 2 at the side 1 in the corner) intersections
            adjacent to the given one. """
        if x[0] == 0:
            l0 = [1]
        elif x[0] == self.size - 1:
            l0 = [self.size - 2]
        else:
            l0 = [x[0] - 1, x[0] + 1]

        if x[1] == 0:
            l1 = [1]
        elif x[1] == self.size - 1:
            l1 = [self.size - 2]
        else:
            l1 = [x[1] - 1, x[1] + 1]

        neighbors_list = []
        for i in l0:
            neighbors_list.append((i, x[1]))
        for j in l1:
            neighbors_list.append((x[0], j))

        return neighbors_list

    def is_occupied(self, x, y):
        return (x, y) in list(self.status.keys())

    def clear(self):
        """ Clear the board """
        self.status = {}
        self.undostack = []
        self.score = {'B': 0, 'W': 0}

    def play(self, pos, color):
        """ This plays a color=black/white stone at pos, if that is a legal move
            and deletes stones captured by that move.
            It returns 1 if the move has been played, 0 if not. """
        if pos in list(self.status.keys()):  # check if empty
            return 0

        if self.legal(pos, color):  # legal move?
            self.status[pos] = color
            captures = self.get_captures(pos, color)
            if captures:
                for x in captures:
                    del self.status[x]   # remove captured stones, if any
                    self.increase_score(color)
            # remember move + captured stones for easy undo
            self.undostack.append((pos, color, captures))
            return captures
        else:
            return 0

    def get_captures(self, pos, color):
        """Returns a list of captured stones resulting
           from placing a color stone at pos """
        c = []  # captured stones

        for x in self.neighbors(pos):
            if x in list(self.status.keys()) and\
                    self.status[x] == self.invert(color):
                c = c + self.hasNoLibExcP(x, self.invert(color), pos)

        if c:
            captures = []
            for x in c:
                if not(x in captures):
                    captures.append(x)
            return captures

        return 0

    def checkKo(self, pos, color):
        ''' Check if a move by color at pos would be a basic Ko infraction '''
        # Basically what we need to check, is if the current play would undo
        # all that was done by the last entry in undostack
        # (capture what was placed and place what was captured).
        if self.undostack:
            lastpos, lastcolor, lastcaptures = self.undostack[-1]
            currentcaptures = self.get_captures(pos, color)
            if lastcaptures != 0 and currentcaptures != 0:
                if lastcolor != color and lastcaptures[0] == pos\
                        and lastpos == currentcaptures[0]:
                    return 1
        return 0

    def legal(self, pos, color):
        """ Check if a play by color at pos would be a legal move. """
        if pos in list(self.status.keys()):
            return 0

        # If the play at pos would leave that stone without liberties,
        # we have two possibilities:
        # 1- It's a capturing move
        # 2- It's an illegal move
        if self.hasNoLibExcP(pos, color):
            # Check if it would capture any stones
            if self.get_captures(pos, color):
                return 1
            # It didnt, so I guess it's illegal
            return 0
        else:
            return not self.checkKo(pos, color)

    def illegal(self, x, y, color):
        """ Check if a play by color at pos would be an illigal move,
            and return pretty errors"""
        if (x, y) in list(self.status.keys()):
            return _('There already is a stone there!')
        if self.checkKo((x, y), color):
            return _('Ko violation!')

        # If the play at pos would leave that stone without liberties,
        # we have two possibilities:
        # 1- It's a capturing move
        # 2- It's an illegal move
        if self.hasNoLibExcP((x, y), color):
            # Check if it would capture any stones
            if self.get_captures((x, y), color):
                return False
            # It didnt, so I guess it's illegal
            return _('Illegal move.')
        else:
            return False

    def hasNoLibExcP(self, pos, color, exc=None):
        """ This function checks if the string (=solidly connected) of stones
            containing the stone at pos has a liberty
            (resp. has a liberty besides that at exc).
            If no liberties are found, a list of all stones in the string is
            returned.

            The algorithm is a non-recursive  implementation of a simple
            flood-filling: starting from the stone at pos, the main while-loop
            looks at the intersections directly adjacent to the stones found so
            far, for liberties or other stones that belong to the string.
            Then it looks at the neighbors of those newly found stones, and so
            on, until it finds a liberty, or until it doesn't find any new
            stones belonging to the string, which means that there are no
            liberties.
            Once a liberty is found, the function returns immediately. """

        # in the end, this list will contain all stones solidly connected to
        # the one at pos, if this string has no liberties in the while loop,
        # we will look at the neighbors of stones in newlyFound
        st = []

        newlyFound = [pos]
        foundNew = 1

        while foundNew:
            foundNew = 0
            # this will contain the stones found in this iteration of the loop
            n = []
            for x in newlyFound:
                for y in self.neighbors(x):
                    if not(y in list(self.status.keys())) and\
                            y != exc and y != pos:    # found a liberty
                        return []
                    # found another stone of same color
                    elif y in list(self.status.keys()) and \
                            self.status[y] == color and \
                            not(y in newlyFound) and not(y in st):
                        n.append(y)
                        foundNew = 1

            st[:0] = newlyFound
            newlyFound = n

        # no liberties found, return list of all stones
        # connected to the original one
        return st

    def get_territories(self):
        def get_group(self, first_element, color):
            current_group = [first_element]
            for current_element in current_group:
                for x in self.neighbors(current_element):
                    if x in list(self.status.keys()):
                        if self.status[x] != color:
                            return None
                    elif not(x in current_group):
                        current_group.append(x)
            return set(current_group)

        groups = {'B': set(), 'W': set()}
        for x in list(self.status.keys()):
            for n in self.neighbors(x):
                if not(n in list(self.status.keys())):
                    new_group = get_group(self, n, self.status[x])
                    if new_group:
                        groups[self.status[x]] |= new_group
        return groups

    def undo(self, no=1):
        """ Undo the last no moves. """
        for i in range(no):
            if self.undostack:
                pos, color, captures = self.undostack.pop()
                del self.status[pos]
                if captures:
                    for p in captures:
                        self.status[p] = self.invert(color)
                        self.score[color] -= 1
                return True
            else:
                return False

    def invert(self, color):
        if color == 'B':
            return 'W'
        else:
            return 'B'

    def get_status(self):
        return self.status


infopanel.py
# -*- coding: utf-8 -*-
# Copyright 2007-2008 One Laptop Per Child
# Copyright 2007 Gerard J. Cerchio <www.circlesoft.com>
# Copyright 2008 Andrés Ambrois <andresambrois@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

from gi.repository import Gtk
from gi.repository import Gdk


class InfoPanel(Gtk.EventBox):

    def __init__(self):
        Gtk.EventBox.__init__(self)

        self.Box = Gtk.VBox()
        self.status_label = Gtk.Label()
        self.Box.pack_start(self.status_label, True, True, 10)

        self.score_label = Gtk.Label()
        self.Box.pack_start(self.score_label, True, True, 10)
        self.modify_bg(Gtk.StateType.NORMAL, Gdk.Color.parse("dark grey")[1])
        self.add(self.Box)

        self.show_all()

    def show(self, text):
        self.status_label.set_text(text)

    def show_score(self, text):
        self.score_label.set_text(text)


activity.py
# -*- coding: utf-8 -*-
# Copyright 2007-2008 One Laptop Per Child
# Copyright 2007 Gerard J. Cerchio <www.circlesoft.com>
# Copyright 2008 Andrés Ambrois <andresambrois@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import logging
import pickle

from gettext import gettext as _
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

from sugar3.activity.activity import Activity
from sugar3.activity.widgets import StopButton
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.graphics.alert import NotifyAlert
from sugar3.graphics.toolbarbox import ToolbarButton

from gametoolbar import GameToolbar
from gogame import GoGame
import boardwidget
import infopanel
from collaboration import CollaborationWrapper
from gtp import gnugo


logger = logging.getLogger('PlayGo')

DEFAULT_SIZE = 19
DEFAULT_KOMI = 5.5


class PlayGo(Activity):

    def __init__(self, handle):
        # Initialize the parent
        Activity.__init__(self, handle)

        logger.debug('Initiating PlayGo')

        self.size = DEFAULT_SIZE
        self.komi = DEFAULT_KOMI

        # Set the activity toolbarbox
        toolbarbox = ToolbarBox()
        self.set_toolbar_box(toolbarbox)

        toolbarbox.toolbar.insert(ActivityToolbarButton(self), -1)
        toolbarbox.toolbar.insert(Gtk.SeparatorToolItem(), -1)

        self.gameToolbar = GameToolbar(self)
        self.gameToolbar.connect('game-restart', self.restart_game)
        self.gameToolbar.connect('game-board-size', self.board_size_change)
        self.gameToolbar.connect('ai-activated', self.ai_activated_cb)
        self.gameToolbar.connect('ai-deactivated', self.ai_deactivated_cb)
        toolbarbox.toolbar.insert(
            ToolbarButton(page=self.gameToolbar, icon_name='txt'), -1)

        separator = Gtk.SeparatorToolItem()
        separator.props.draw = False
        separator.set_expand(True)
        toolbarbox.toolbar.insert(separator, -1)

        toolbarbox.toolbar.insert(StopButton(self), -1)

        toolbarbox.show_all()
        toolbarbox.toolbar.show_all()
        self.gameToolbar.show_all()

        # Initialize the game
        self.game = GoGame(self.size)
        self.CurrentColor = 'B'
        self.PlayerColor = 'B'
        self.pass_count = 0
        self.ai_activated = False
        self.set_up_ui()

        if not handle.object_id:
            self.infopanel.show(_('Welcome to PlayGo!'))
        else:
            self.show_score()

        self.lastX = -1
        self.lastY = -1

        # Set keypad actions
        self._key_actions = {
            'KP_Up': 'move_up',
            'KP_Right': 'move_right',
            'KP_Down': 'move_down',
            'KP_Left': 'move_left',
            'KP_Home': 'place_stone',
            'KP_Next': 'undo',
            'KP_End': 'pass'}

        # Set up collaboration
        self.collaboration = CollaborationWrapper(
            self, self.buddy_joined, self.buddy_left, self.Play,
            self.game.undostack, self.bootstrap)

        self.connect('shared', self.collaboration._shared_cb)
        if self.get_shared_activity():
            # We are joining the activity
            self.connect('joined', self.collaboration._joined_cb)
            if self.get_shared():
                # We've already joined
                self.collaboration._joined_cb()

    def set_up_ui(self):
        self.board = boardwidget.GoBoardWidget(
            self.game.get_status(), self.size)
        self.board.connect('motion-notify-event', self.board_motion_cb)
        self.board.connect('insert-requested', self.insert_cb)

        self.main_view = Gtk.VBox()

        self.board_aspect = Gtk.AspectFrame.new(None, .5, .5, 1, False)
        self.board_aspect.add(self.board)
        self.main_view.pack_start(self.board_aspect, True, True, 0)

        self.buttons_box = Gtk.HBox()
        self.buttons_alignment = Gtk.Alignment.new(0.5, 1, 0.5, 1)
        # Pass button
        self.pass_button = Gtk.Button(_('Pass'))
        self.pass_button.connect("clicked", self.pass_cb)
        self.buttons_box.pack_start(self.pass_button, True, True, 10)

        # Undo button
        self.undo_button = Gtk.Button(_('Undo'))
        self.undo_button.connect("clicked", self.undo_cb)
        self.buttons_box.pack_start(self.undo_button, True, True, 10)

        self.buttons_alignment.add(self.buttons_box)
        self.main_view.pack_start(self.buttons_alignment, False, True, 10)

        self.infopanel = infopanel.InfoPanel()
        self.main_view.pack_start(self.infopanel, False, True, 0)

        self.set_canvas(self.main_view)
        self.show_all()

    def insert_cb(self, widget, x, y, announce=True, ai_play=False):
        '''The insert function. It makes the play and manages turn changing
           stone drawing, etc.

        Parameters x and y are the coordinates of the play ((0,0) is top left),
        widget points to the widget that emitted the signal connected to this
        function, announce is True when we need to announce this play to
        other people collaborating, and ai_play is True when this is called
        by the AI, so we know not to ask for an AI play again'''

        # Check if it's our turn only if it's a local play (announce is True)
        # Calls by other players will always be out of turn for us.
        if announce and self.get_currentcolor() != self.get_playercolor():
            logger.debug('Play at %s x %s was out-of-turn!', x, y)
            self.infopanel.show(_('It\'s not your turn!'))
            return False
        # Make the play only if it wasn't a pass move.
        if x != -1:
            self.pass_count = 0
            error = self.game.illegal(x, y, self.get_currentcolor())
            if error:
                self.infopanel.show(error)
                return False
            # Make the play
            captures = self.game.play((x, y), self.get_currentcolor())
            if self.ai_activated and not ai_play:
                self.notify_ai(x, y, self.get_currentcolor())

            self.gameToolbar.grey_out_size_change()

            if captures:
                self.redraw_captures(captures)

            self.show_score()
            self.board.draw_stone(
                self.board.context, x, y, self.get_currentcolor(), widget)
            self.board.queue_draw()
        # Player passed
        else:
            self.pass_count += 1
        # Announce the local play
        if self.get_shared() and announce:
            self.collaboration.Play(x, y)
        self.change_turn()
        if x == -1:
            self.infopanel.show(_('Opponent passed'))
        # If this is the second consecutive pass, the game ends
        if self.pass_count == 2:
            self.game_end()
            return
        # If we are playing a local game with AI turned off, change the color
        if not self.get_shared() and not self.ai_activated:
            self.change_player_color()
        # Else, if the AI is on, and this wasn't played by it,
        # request a play by it.
        elif self.ai_activated:
            self.change_player_color()
            if not ai_play:
                self.play_ai()

    def undo_cb(self, widget=None, data=None):
        if self.game.undo():
            self.board.queue_draw()
            # If playing against AI undo twice
            if self.ai_activated:
                self.ai.undo()
                self.game.undo()
                self.ai.undo()
            else:
                self.change_turn()

            if not self.get_shared() and not self.ai_activated:
                self.change_player_color()

            self.show_score()

    def pass_cb(self, widget=None, data=None):
        if self.get_shared():
            if self.get_currentcolor() == self.get_playercolor():
                self.pass_count += 1
                self.collaboration.Play(-1, -1)
            else:
                self.infopanel.show(_('It\'s not your turn!'))
                return
        else:
            self.pass_count += 1
            self.change_player_color()

        self.change_turn()

        if self.pass_count == 2:
            self.game_end()

        if self.ai_activated:
            self.ai.pass_move(self.get_currentcolor())
            self.play_ai()

    def write_file(self, file_path):
        logger.debug('Writing file: %s', file_path)
        # Strip the undostack
        undostack = self.game.undostack[:]
        strippedstack = []
        for pos, color, captures in undostack:
            strippedstack.append(pos)

        f = open(file_path, 'wb')

        try:
            pickle.dump(strippedstack, f, pickle.HIGHEST_PROTOCOL)
        except Exception as e:
            print("Error ", e, " occured")
        finally:
            f.close()

        self.metadata['our-color'] = self.get_playercolor()
        self.metadata['shared'] = str(self.get_shared())
        self.metadata['size'] = str(self.size)
        self.metadata['ai'] = str(self.ai_activated)

    def read_file(self, file_path):
        logger.debug('Reading file: %s', file_path)
        f = open(file_path, 'r')
        try:
            newstack = pickle.load(f)
        finally:
            f.close()

        if self.get_shared():
            logger.debug('The game we are loading is shared!')
            self.PlayerColor = self.metadata.get('our-color', 'B')
        if self.size != self.metadata.get('size', DEFAULT_SIZE):
            self.board_size_change(
                None, int(self.metadata.get('size', DEFAULT_SIZE)))

        self.bootstrap(newstack)

        if self.metadata.get('ai', None) == 'True':
            self.gameToolbar._size_combo.set_sensitive(False)
            # This triggers the 'toggled' signal too
            self.gameToolbar._ai_button.set_active(True)

    def board_motion_cb(self, widget, event):
        x, y = self.board.get_mouse_event_xy(event)
        if x == self.lastX and y == self.lastY:
            return

        if not self.game.is_occupied(x, y)\
                and self.game.legal((x, y), self.get_playercolor()):
            self.board.draw_ghost_stone(x, y, self.get_playercolor())
            self.lastX = x
            self.lastY = y

    def invert_color(self, color):
        if color == 'B':
            return 'W'
        return 'B'

    def get_currentcolor(self):
        return self.CurrentColor

    def change_turn(self):
        # It's the other guy's turn now
        if self.CurrentColor == 'B':
            self.infopanel.show(_('White\'s turn'))
        else:
            self.infopanel.show(_('Black\'s turn'))

        self.CurrentColor = self.invert_color(self.get_currentcolor())

    def get_playercolor(self):
        return self.PlayerColor

    def change_player_color(self):
        self.PlayerColor = self.invert_color(self.get_playercolor())

    def set_player_color(self, color):
        self.PlayerColor = color

    def redraw_captures(self, captures):
        for x in captures:
            self.board.redraw_area(x[0], x[1])

    def bootstrap(self, plays):
        ''' Take our game to the state it would have if
            @plays were manually played'''
        logger.debug('Bootstraping...')
        # HACK: Looks like read_file is called before the board is exposed
        self.board.queue_draw()
        for pos in plays:
            logger.debug(
                'Playing at %s with color %s', pos, self.get_currentcolor())
            captures = self.game.play(
                (pos[0], pos[1]), self.get_currentcolor())
            if captures:
                self.redraw_captures(captures)

            self.change_turn()
            self.change_player_color()

        logger.debug('Color after bootstraping is %s', self.get_currentcolor())
        self.show_score()
        self.board.queue_draw()

    def restart_game(self, widget=None):
        logger.debug('Received restart signal!')
        self.CurrentColor = 'B'
        self.PlayerColor = 'B'
        self.pass_count = 0
        self.game.clear()
        self.board.territories = None
        self.board.status = self.game.status
        self.board.queue_draw()
        self.show_score()
        self.board.set_sensitive(True)
        self.buttons_box.set_sensitive(True)
        self.lastX = -1
        self.lastY = -1
        if self.ai_activated:
            self.ai.clear()

    def game_end(self):
        self.board.set_sensitive(False)
        self.buttons_box.set_sensitive(False)
        territories = self.game.get_territories()
        self.board.territories = territories

        final_score = {
            'B': (len(territories['B']) - self.game.get_score()['W']),
            'W': (
                len(territories['W']) -
                self.game.get_score()['B'] + self.komi)}

        if final_score['B'] > final_score['W']:
            winner_string = _('Black wins!')

        elif final_score['W'] > final_score['B']:
            winner_string = _('White wins!')

        else:
            winner_string = _('There was a tie!')

        self.infopanel.show(_('Game ended! %s' % winner_string))
        self.infopanel.show_score(
            _('Final score: White %(W)g - Black %(B)g' % final_score))

    def board_size_change(self, widget, size):
        self.lastY = -1
        self.lastX = -1
        if size == self.size:
            return

        self.size = size
        del self.game
        self.game = GoGame(size)
        self.board_aspect.remove(self.board)
        del self.board
        self.board = boardwidget.GoBoardWidget(
            self.game.get_status(), int(size))
        self.board_aspect.add(self.board)
        self.board.connect('motion-notify-event', self.board_motion_cb)
        self.board.connect('insert-requested', self.insert_cb)
        self.board.show()
        if self.ai_activated:
            del self.ai
            self.ai = gnugo(boardsize=self.size)

    def ai_activated_cb(self, widget=None):
        try:
            self.ai = gnugo(boardsize=self.size)
        except Exception as e:
            self._alert(_('AI'), _('GnuGo loading failed!: %s' % e))
            self.gameToolbar.set_ai_button_state(False)

        else:
            self.ai_activated = True
            for pos, color, captures in self.game.undostack:
                self.notify_ai(pos[0], pos[1], color)

            self._alert(_('AI'), _('PlayGo AI Activated'))

    def ai_deactivated_cb(self, widget):
        if self.ai:
            self.ai_activated = False
            del self.ai
            self._alert(_('AI'), _('PlayGo AI Deactivated'))

    def notify_ai(self, x, y, color):
        logger.debug('Notifying AI of play by %s at %s x %s', color, x, y)
        self.ai.make_play(color, x, y)

    def play_ai(self):
        if self.get_currentcolor() == self.get_playercolor():
            x, y = self.ai.get_move(self.get_currentcolor())
            logger.debug('Got play %s x %s from AI', x, y)
            self.insert_cb(None, x, y, ai_play=True)

    def show_score(self):
        self.infopanel.show_score(
            _("Score is: White %(W)g - Black %(B)g" % self.game.get_score()))

    def _alert(self, title, text=None):
        alert = NotifyAlert(timeout=5)
        alert.props.title = title
        alert.props.msg = text
        self.add_alert(alert)
        alert.connect('response', self._alert_cancel_cb)
        alert.show()

    def _alert_cancel_cb(self, alert, response_id):
        self.remove_alert(alert)

    # ----------- Keypad events functions ----------- #
    def _key_pressed_cb(self, grabber, keycode, state):
        key = grabber.get_key(keycode, state)
        logging.debug("Key pressed: %s", key)
        action = self._key_actions[key]
        method = getattr(self, 'handle_' + action)
        method()

    def handle_move_up(self):
        self.move_ghost_stone(0, -1)

    def handle_move_right(self):
        self.move_ghost_stone(1, 0)

    def handle_move_down(self):
        self.move_ghost_stone(0, 1)

    def handle_move_left(self):
        self.move_ghost_stone(-1, 0)

    def handle_place_stone(self):
        self.insert_cb(None, self.lastX, self.lastY)

    def handle_undo(self):
        self.undo_cb()

    def handle_pass(self):
        self.pass_cb()

    def move_ghost_stone(self, deltaX=0, deltaY=0):
        if self.lastX < 0:
            self.lastX = int(self.size / 2)
            self.lastY = int(self.size / 2)
        x = self.lastX + deltaX
        y = self.lastY + deltaY
        if x < 0 or x > self.size - 1 or y < 0 or y > self.size - 1:
            return

        self.board.draw_ghost_stone(x, y, self.get_playercolor())
        self.lastX = x
        self.lastY = y

    # ------- Callbacks for Collaboration ------- #
    def buddy_joined(self, buddy):
        self._alert(_('Buddy joined'), _('%s joined' % buddy.props.nick))

    def buddy_left(self, buddy):
        self._alert(_('Buddy left'), _('%s left' % buddy.props.nick))

    def Play(self, x, y, sender=None):
        ''' Called when a stone was placed at x,y by sender'''
        # Discard a pass move received in our turn.
        # Do it here for extra security
        if x == -1 and self.get_currentcolor() == self.get_playercolor():
            return
        self.insert_cb(None, x, y, False)
